### PR TITLE
Enable to build PDF docs on Ubuntu precise

### PR DESCRIPTION
**Pull request by tk0miya**
I try to build PDF docs on Ubuntu precise (through Travis CI).
But current stable branch has some troubles to build it.
This resolves the errors.

At same time, this enables to build continuously on Travis CI.

@jfbu Could you review this please?

### CODE REVIEW

**Comment by jfbu** [comment]
I will go during next few hours check on a place where I have Ubuntu 12.04. I don't have admin rights there, and use a  TeXLive 2015 install in my user space, but as I wanted to install recently released TL2016 there, I can profit of this occasion to revert temporarily to the naked situation from my sysadmin (I forgot which release of TeXLive is there by default; besides the file system is really complicated and it might be not the one native from Ubuntu).
- Strange that you need to ship framed.sty, (it exists since 1992 and was last updated on 2011/10/22), testing on TL2012 which is the oldest I have on my laptop I do get:

```
$ tlmgr info framed
package:     framed
category:    Package
shortdesc:   Framed or shaded regions that can break across pages.
longdesc:    The package creates three environments: - framed, which puts an ordinary frame box around the region, - shaded, which shades the region, and - leftbar, which places a line at the left side. The environments allow a break at their start (the \FrameCommand enables creation of a title that is "attached" to the environment); breaks are also allowed in the course of the framed/shaded matter. There is also a command \MakeFramed to make your own framed-style environments.
installed:   Yes
revision:    26789
cat-version: 0.96
cat-date:    2012-06-01 12:48:04 +0200
cat-license: other-free
collection:  collection-latexextra
```

hence it was (and still is) in collection-latexextra which you have added to travis.yml. Perhaps Ubuntu 12.04 ships something like TL2010, but \OuterFrameSep was not yet in that version of framed.sty. Will check later.
- regarding use of `\c@chapter` looks good to me; perhaps do test as
  
  ``` latex
  \ifdefined\c@chapter
     ...
  \else
     ...
  \fi
  ```
- the use of newfloat package for literal-block may need discussion (too late for 1.4.4 I guess). 
  - the latex writer does not seem to write any `literal-block` environment
  - literal-blocks are in fact not treated as LaTeX floats by Sphinx as they
    are handled by custom Verbatim,
  - only \caption needs to be functional to set up "Listing ..."
  - as `literal-block` has a dash, user can not do `\listofliteral-blocks` as `\listoffigures` and `\listoftables` but must do `\csname listofliteral-blocks\endcsname` which thus appears as an undocumented possibility
    
    ``` rst
    Hello
    =====
    
    .. raw:: latex
    
      \csname listofliteral-blocks\endcsname
      \relax
    
    Section
    -------
    
    .. code-block:: python
      :caption: TEST
    
      coucou
    ```
  - package newfloat has lots of code to test for special document classes,
    which are not Sphinx related, with the view to make things like
    `\listof..` work, but as said above this feature does not look in case of
    literal-block to be made accessible to Sphinx users.
  - perhaps Sphinx originally used
    
    ``` latex
    \begin{literal-block}
    ...
    \caption{my code}
    \end{literal-block}
    ```
    
    which is made possible by the `\DeclareFloatingEnvironment{literal-block}`etc.. commands at bottom of sphinx.sty. But nowadays, as this is not syntax used by Sphinx latex writer, we have no need for newfloat, and could do the counter definitions (including reset when section or chapter changes) and caption definition (with Listing) with a few lines of direct basic LaTeX code (I need to confirm this, but this is my belief).
    I will look today, but probably too late for 1.4.4 to drop newfloat ?

I will report on experiment on the Ubuntu 12.04 I have access to, in a few
hours.

**Comment by tk0miya** [comment]
I will work for 1.4.4 on 6/12 during 08:00-10:00 UTC.
But it's okay to postpone if you need time to review.
So please review this on your pace :-)

**Comment by tk0miya** [comment]
Memo: Precise uses TeXLive-2009:
http://packages.ubuntu.com/precise/texlive-latex-extra

**Comment by tk0miya** [comment]
The `literal-block` environment was introduced at 447c8c1d377374e59f3a2a2549fa400de486f4f8 (v1.3). Originally, it is used to represent a caption of code-block using `\caption` macro.

The current structure appeared at a1240d2c. It removes `literal-block` macro to allow page breaks. So these code is only used to represent captions. It would be nice if we can drop dependency to `newfloat.sty`.

But I think it's not a bug fix, so master branch is much better to fix it.

**Comment by jfbu** [comment]
Ok, testing now. Removing from PATH my own install (TL2015) and the sysadmin one (TL2013) I do get as you say TeXLive-2009 and I do confirm the problem with the framed.sty lacking `\OuterFrameSep`. 

As an alternative to shipping framed.sty, the following could be done:

```
diff --git a/sphinx/texinputs/sphinx.sty b/sphinx/texinputs/sphinx.sty
index 4499247..12ec1ad 100644
--- a/sphinx/texinputs/sphinx.sty
+++ b/sphinx/texinputs/sphinx.sty
@@ -416,7 +416,7 @@
 \newenvironment{SphinxShadowBox}
   {\def\FrameCommand {\Sphinx@ShadowFBox }%
    % configure framed.sty not to add extra vertical spacing
-   \OuterFrameSep \z@skip
+   \ifdefined\OuterFrameSep \OuterFrameSep\z@skip \fi
    % the \trivlist will add the vertical spacing on top and bottom which is
    % typical of center environment as used in Sphinx <= 1.4.1
    % the \noindent has the effet of an extra blank line on top, to
@@ -532,7 +532,7 @@
    % configure framed.sty's parameters to obtain same vertical spacing
    % as for "light" boxes. We need for this to manually insert parskip glue and
    % revert a skip done by framed before the frame.
-    \setlength{\OuterFrameSep}{0pt}
+    \ifdefined\OuterFrameSep \OuterFrameSep\z@skip \fi
     \vspace{\FrameHeightAdjust}
    % copied/adapted from framed.sty's snugshade
    \def\FrameCommand##1{\hskip\@totalleftmargin

```

With it make latexpdf succeeds on Ubuntu 12.04 with TeXLive 2009/Debian.

```
This is pdfTeX, Version 3.1415926-1.40.10 (TeX Live 2009/Debian)
entering extended mode
(./sphinx.tex
LaTeX2e <2009/09/24>
Babel <v3.8l> and hyphenation patterns for english, usenglishmax, dumylang, noh
...
Output written on sphinx.pdf (276 pages, 2120740 bytes).
Transcript written on sphinx.log.
```

I have tested it with a few files using captioned literal-blocks, seemed okay.

Now, I see a problem on TL2009/Debian that there is too much vertical space if a captioned literal block appears in a list. I have not investigated where this comes from, and using the more recent framed.sty does not change the issue. Should be related to some package affecting \caption or \abovecaptionskip possibly. I don't see the issue when using TeXLive 2013.

```
#. code at level 1

   .. code-block:: python
      :caption: NIVEAU 1

      1+1

   #. code at level 2

      .. code-block:: python
         :caption: NIVEAU 2

         1+1

      #. code at level 3

         .. code-block:: python
            :caption: NIVEAU 3

            1+1

         #. code at level 4

            .. code-block:: python
               :caption: NIVEAU 4

               1+1
```

Perhaps some package influences \abovecaptionskip, anyway I did not investigate further as this would take time and there is no issue with TL2013 (I will test on TL2012 later).

As per the problem with `\thechapter` I do not see it happen here on Precise+TL2009/Debian. Anyway, it is worthwile to do for the time being in sphinx.sty (diff continued from above)

```
@@ -841,10 +841,10 @@
 % Define literal-block environment
 \RequirePackage{newfloat}
 \DeclareFloatingEnvironment{literal-block}
-\ifx\thechapter\undefined
-  \SetupFloatingEnvironment{literal-block}{within=section,placement=h}
-\else
+\ifdefined\c@chapter
   \SetupFloatingEnvironment{literal-block}{within=chapter,placement=h}
+\else
+  \SetupFloatingEnvironment{literal-block}{within=section,placement=h}
 \fi
 \SetupFloatingEnvironment{literal-block}{name=List}
 % control caption around literal-block
```

at bottom of sphinx.sty.

As I explained before, as far as I can tell Sphinx never uses the `literal-block` float environment defined with help of `newfloat` hence we could possibly avoid the code above, replace it by some custom code to defined literal-block as counter usable with `\caption` and then this would avoid having to ship newfloat.sty. i will examine that today June 11...

**Comment by jfbu** [comment]
@tk0miya I missed your comment appearing while I had started looking at Precise situation. Thanks for identifying the commits and explanations. I agree better removing newfloat dependency on master branch. Will look into it later today.

File: CHANGES [hunk 1]
```diff
@@ -17,6 +17,7 @@ Bugs fixed
 * latex: allow one more nesting depth for code-blocks (ref #777, PR #2624)
 * #2635: Latex code directives produce inconsistent frames based on viewing resolution
 * #2639: Sphinx now bundles iftex.sty
+* Sphinx now bundles framed.sty
```
Comment (jfbu) [inline CHANGES]: @tk0miya shouldn't this line be suppressed if we use \OuterFrameSep only conditionally? no need to ship framed.sty then. Memo: version in TeXLive 2009 of framed.sty is `v 0.95    2007/10/04`, the one current is `v 0.96    2011/10/22` hence shipped with TL-2012, or up-to-date TL-2011.

**Comment by tk0miya** [comment]
Thank you for advice! I updated sphinx.sty and dropped `framed.sty`.

**Comment by jfbu** [comment]
ok, LGTM.

side note: I tested using TeXLive 2012 + old framed.sty as in TeXLive 2009 and can not reproduce the issue with too much vertical space above captions of literal-block in lists. On the other hand with TeXLive 2009/Debian + new current framed.sty, there is the issue. So, whatever the cause, it does not seem to be related to use of framed.sty.

File: CHANGES (hunk 1 above)
Comment (tk0miya) [inline CHANGES]: Oh, I've forgotten to update CHANGES.

**Comment by jfbu** [comment]
I have identified the cause of the problem on Ubuntu Precise (TeXLive 2009/Debian) creating big spaces above captions of literal blocks in lists: it is due to faulty needspace.sty. When I use current needspace.sty `2010/09/12 v1.3d` the problem disappears. The one on TL2009/Debian is `2009/09/03 v1.3b`. Perhaps sphinx.sty should do `\RequirePackage{needspace}[2010/09/12]`, but this will again break your build on Precise... I will check if simple correction is possible, short of shipping correct needspace.sty.

**Comment by jfbu** [comment]
@tk0miya I think we can ship current needspace.sty. It is very short package and shipping it is not much more than correcting the bug from 2009/09/03 v1.3b as  in current 2010/09/12 v1.3d. Also the current version is stripped of comments, hence only about 35 lines. This is only needed for people using such antequated TeX as TeXLive 2009, but not much of added weight to Sphinx.

**Comment by tk0miya** [comment]
Ok, I will add it later.
Thank you for very helpful reviewing!

**Comment by tk0miya** [comment]
Added needspace.sty.

**Comment by jfbu** [comment]
@tk0miya fwiw I have determined the cause of the problem with `\thechapter` on Ubuntu Precise, i.e. TeXLive 2009/Debian. The problem is due to a bug in hyperref.sty 2009/10/09 v6.79a (I didn't see it yesterday as I actually did not  try `sphinxhowto` class). On line 4776 of this old hyperef, it does

```
\@ifundefined{thechapter}{%
```

but it is well known after effect of LaTeX kernel command `\@ifundefined` that the tested control sequence then becomes `\relax` if it was not defined before. The test `\ifx\thechapter\undefined` then fails to say that `\thechapter` is undefined, because actually `\thechapter` has acquired the `\relax` meaning, hence it is not undefined. However doing `\@ifundefined{thechapter}{..}{..}` would have been correct, because `\@ifundefined` chooses the true branch in case the macro has meaning `\relax`. In short, once it has been used, this LaTeX command forces all followers to use it again.

In hyperref 2012/11/06 v6.83m which is thus part of any uptodate TeXLive 2012, this has been corrected to 

```
\ltx@IfUndefined{thechapter}{%
```

where `\ltx@IfUndefined` comes from package `ltxcmds.sty`, loaded by hyperref. I did not try to check which version of hyperref does that first. The `\ltx@IfUndefined` is an improved version of the LaTeX kernel `\@ifundefined` which uses an extra scope limiting group to avoid the after-effect above. Then `\thechapter` remains undefined after the test, and not `\relax`.

Memo: the problem with `\ifx\thechapter\undefined` possibly being fooled applies also to `\ifdefined\thechapter`, in case `\thechapter` has somehow acquired `\relax` meaning, this test will consider `\thechapter` to be defined. I recommended `\ifdefined \OuterFrameSep`. I consider it is very unlikely some package does `\@ifundefined{OuterFrameSep}{..}{..}` earlier, hence I am not too worried.

But for the future we should keep in mind all `\ifx\foo\undefined` or `\ifdefined\foo` tests may be fooled by use of `\@ifundefined` (not protected by a group) by some earlier package. On the other hand using `\@ifundefined` ourselves would be poor coding style due to its aftereffect.

**Comment by tk0miya** [comment]
Okay, let's discuss in #2660  how to check what sectioning unit is used.
I believe `\ifdefined \OuterFrameSep` is not related with sectioning unit.

**Pull request merged**